### PR TITLE
feat(r-lang): Add module for R programming language

### PR DESCRIPTION
**Pull request by jimleroyer**
#### Description

Added a module for the R programming language. 

#### Motivation and Context

R is number 10 on the latest [Tiobe index](https://www.tiobe.com/tiobe-index/). It made its mark as a statistical language. Putting on support for most popular languages among developers makes starship a bit more attractive with its support arsenal for commonly used tools.

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Screenshots (if appropriate):

image

#### How Has This Been Tested?

- [ ] I have tested using **MacOS**
- [ ] I have tested using **Linux**
- [X] I have tested using **Windows**

#### Checklist:

- [X] I have updated the documentation accordingly.
- [X] I have updated the tests accordingly.

### CODE REVIEW

File: docs/config/README.md
Comment (vladimyr) [inline docs/config/README.md:1737]: I believe you should check for `.Rproj` files too.

Comment (jimleroyer) [reply docs/config/README.md:1920]: Just added this file type. Thanks for the review and sorry for the delay!

**Comment by vladimyr** [comment]
@jimleroyer Is this ready for review?

**Comment by jimleroyer** [comment]
@vladimyr It can be, after the errors are fixed. I couldn't figure why it complains that a trait isn't properly implemented. I would need some help on that. We can chat in Discord further more to fix this.

File: src/configs/r.rs
Comment (davidkna) [inline src/configs/r.rs:14]: This is implemented like this now:
```suggestion
impl<'a> Default for RConfig<'a> {
    fn default() -> Self {
```

Also can please you add the module to `FullConfig` in `configs/mod.rs`?

Comment (jimleroyer) [reply src/configs/r.rs:14]: Ok my mistake, I just realized I didn't merge the starship repository with my fork and got out of sync. Thanks for the quick response. Let me take another look to the PR with latest changes then. 👍

File: src/configs/r.rs
Comment (vladimyr) [inline src/configs/r.rs:12]: This lacks `detect_extensions`, `detect_files` and `detect_folders` members.

File: src/modules/r.rs
Comment (vladimyr) [inline src/modules/r.rs:27]: Version extraction and formatting should be done lazily, please move it under `version` arm of the formatter variable mapper.

File: src/modules/r.rs
Comment (vladimyr) [inline src/modules/r.rs:76]: I think you have a typo here:
```suggestion
        let r_v3 = r#"R version 3.6.3 (2020-02-29) -- "Holding the Windsock"
```

Comment (vladimyr) [reply src/modules/r.rs:76]: This also needs to get added to the global mocks dictionary located inside `src/utils.rs` as part of `exec_cmd` implementation.

File: src/modules/r.rs
Comment (vladimyr) [inline src/modules/r.rs:89]: This can be removed or rather replaced with tests checking prompt rendering inside a folder with target files/extensions/folders. See other modules for getting an idea of what kind of tests are needed here.

File: src/modules/r.rs
Comment (vladimyr) [inline src/modules/r.rs:63]: This can be done without using regex:
```suggestion
fn parse_version(r_version: &str) -> Option<String> {
    let version = r_version
        .lines()
        // take first line
        .next()
        // split into ["R", "version", "3.6.3", "(2020-02-29)", ...]
        .split_whitespace()
        // return "3.6.3"
        .nth(2)?;

    Some(format!("v{}", r_version))
}
```

File: src/modules/r.rs
Comment (vladimyr) [inline src/modules/r.rs:13]: Please update this docstring to list all conditions.

File: src/modules/r.rs
Comment (vladimyr) [inline src/modules/r.rs:1]: Go's module is named `src/modules/golang.rs` while config is named `src/configs/go.rs`. If we apply that here we get `src/modules/rlang.rs` and the corresponding config becomes `src/config/r.rs`.

**Review by vladimyr — CHANGES_REQUESTED** [review]

File: docs/config/README.md
Comment (jimleroyer) [reply docs/config/README.md:2251]: Sure, let's add these too.

File: src/modules/r.rs
Comment (jimleroyer) [reply src/modules/r.rs:1]: Hmm I think Go is popularly known as Go lang, whereas other modules are not called pythonlang, rustlang, etc... I admit though that `r` is quite a bit short and cryptic. Single letter language could have this suffix such as dlang, clang, etc... What is your reasoning? is this align on your thinking?

File: src/modules/r.rs
Comment (jimleroyer) [reply src/modules/r.rs:94]: Ok I will check this tomorrow.

File: src/modules/r.rs
Comment (jimleroyer) [reply src/modules/r.rs]: I had to modify this code a bit, if you want to take a second look.

File: src/modules/r.rs
Comment (vladimyr) [reply src/modules/r.rs]: > I admit though that `r` is quite a bit short and cryptic. Single letter language could have this suffix such as dlang, clang, etc...

You've pretty much said it all - adding lang suffix to avoid too short names following golang's precedent.

File: docs/config/README.md
Comment (Milo123459) [inline docs/config/README.md:2255]:
```suggestion
- The current directory contains a file with `.r` extension
```

Comment (Milo123459) [reply docs/config/README.md:2255]: ^ Purely for consistency as all other modules use a similar grammatical system like that

Comment (jimleroyer) [reply docs/config/README.md]: Thank you for the suggestion.

File: src/modules/r.rs
Comment (jimleroyer) [reply src/modules/r.rs]: Module name has been changed to `rlang`.

File: src/configs/rlang.rs
Comment (Milo123459) [inline src/configs/rlang.rs:24]: I think you should make these lowercase. I believe it should work for any case?

Comment (Milo123459) [reply src/configs/rlang.rs:24]: ^^ same with all others

Comment (jimleroyer) [reply src/configs/rlang.rs:24]: Can you tell me your thinking for it to be lowercase? is that a convention that starship has?

For correctness and if possible, I would keep it as-is, unless another reason supplants it. The `R` file extensions are normally upper case for R source code (not enforced but the standard), either from its own [source code](https://svn.r-project.org/R/trunk/src/library/base/R/), [its libraries](https://github.com/tidyverse/tidyverse) or by files generated by tools such as RStudio (the official IDE).

Comment (Milo123459) [reply src/configs/rlang.rs:24]: That's just how I do it and most other modules do it like that so for consistency I suppose

File: src/modules/rlang.rs
Comment (davidkna) [inline src/modules/rlang.rs:31]: Can you add version formatting to this module? #2611

File: src/configs/rlang.rs
Comment (davidkna) [reply src/configs/rlang.rs:24]: I think correctness is more important than consistency here.

**Review by davidkna — COMMENTED** [review]
Beyond this LGTM

File: src/configs/rlang.rs
Comment (davidkna) [reply src/configs/rlang.rs:24]: I think correctness is more important than consistency here.

File: src/modules/rlang.rs
Comment (jimleroyer) [reply src/modules/rlang.rs]: Just added it!

File: src/modules/rlang.rs
Comment (vladimyr) [inline src/modules/rlang.rs:66]: Use `?` here so you can simply invoke `split_whitespace` instead of chaining with `and_then`. It is more readable and in line with version extractors present in other modules.

File: src/modules/rlang.rs
Comment (vladimyr) [inline src/modules/rlang.rs:90]: Unless it is indented in actual command output this shouldn't look like this. In this particular case, it is irrelevant but we should always strive to faithfully replicate stuff regardless.

File: src/modules/rlang.rs
Comment (vladimyr) [inline src/modules/rlang.rs:98]: I know it is tedious and frankly a boring job and that we should build a macro to alleviate the pain but we really should test **all** patterns not just `*.R`. So please do the c/p here.

File: src/utils.rs
Comment (vladimyr) [inline src/utils.rs:177]: Hmm, this looks different from the text you used for the version extraction test?
Also, use r# string to get rid of escapes and extra? line feeds.

**Review by vladimyr — CHANGES_REQUESTED** [review]

File: docs/config/README.md
Comment (davidkna) [inline docs/config/README.md:2289]: Please also add `version_format` to the docs.

```suggestion
| `format`            | `"via [$symbol($version )]($style)"` | The format for the module.                    |
| `version_format` | `"v${raw}"` | The version format. Available vars are `raw`, `major`, `minor`, & `patch`|
```

File: src/utils.rs
Comment (jimleroyer) [reply src/utils.rs:177]: Done..

File: src/modules/rlang.rs
Comment (jimleroyer) [reply src/modules/rlang.rs:98]: Done..

**Comment by matchai** [comment]
It looks like we're still running into lifetime issues in nightly before it can be merged, if you've got the chance to look into it. 😊

**Review by davidkna — APPROVED** [review]
LGTM

**Review by matchai — APPROVED** [review]

**Comment by matchai** [comment]
Thank you for the PR, Jimmy, and everyone else for the reviews! 😊

**Comment by mjsteinbaugh** [comment]
@jimleroyer This isn't picking up my R version correctly on macOS. Anything I need to do to configure?

**Comment by jimleroyer** [comment]
@mjsteinbaugh Indeed, a quick test on my MacBook tells me you are right: it does not work on maxOS, yet it does on Windows.

I tested further for a quick hunch: the `R --version` command on macOS redirects to the standard output while the Windows version redirects to the standard error. I developed this feature on my Windows machine hence, the [code catches the standard error](https://github.com/starship/starship/pull/1475/files#diff-829ec5fc815ce89e5c672dd750c8ea1ccb045e80d977eb55ee3e3f090b55fa3cR58). We'll need something to accommodate both platform.

I will open a bug tonight and see what I can do in the next days to fix it.

**Pull request merged**
